### PR TITLE
openssl rehash: add check for OPENSSL_strdup

### PR DESCRIPTION
**Pull request by JiangJias**
As the potential failure of the memory allocation,
it should be better to check the return value of
OPENSSL_strdup() and return error if fails.
Also, we need to restore the 'ep' to be NULL if fails.
And since the OPENSSL_strdup() will print report if
fails, the BIO_print() is not needed.

Signed-off-by: Jiasheng Jiang <jiasheng@iscas.ac.cn>

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: apps/rehash.c [hunk 1]
```diff
@@ -168,6 +168,11 @@ static int add_entry(enum Type type, unsigned int hash, const char *filename,
         *ep = nilhentry;
         ep->old_id = ~0;
         ep->filename = OPENSSL_strdup(filename);
+        if (ep->filename == NULL) {
+            OPENSSL_free(ep);
```
Comment (t8m) [inline apps/rehash.c:172]: Perhaps some BIO_printf with an error message would be a good idea?

Comment (JiangJias) [reply apps/rehash.c:172]: OK, I have amended commit message and submitted a fixup.

**Review by paulidale — APPROVED** [review]
I think that there is a case for an app_strdup() function that handles failures like app_malloc().

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you for your contribution.

**Pull request closed without merging**